rasa/utils/train_utils.py

```diff
@@ -534,6 +705,71 @@ def sample_negatives(
     )
 
 
+def tf_matmul_sparse(inputs: tf.SparseTensor, kernel: tf.Tensor):
+    def map_function(x):
+        i, dense_slice = x[0], x[1]
+        sparse_slice = tf.sparse.reshape(
+            tf.sparse.slice(
+                inputs, [i, 0, 0], [1, inputs.dense_shape[1], inputs.dense_shape[2]]
+            ),
+            [inputs.dense_shape[1], inputs.dense_shape[2]],
+        )
+        mult_slice = tf.sparse.matmul(sparse_slice, dense_slice)
+        return mult_slice
+
+    elems = (tf.range(0, inputs.dense_shape[0], delta=1, dtype=tf.int64), kernel)
+    return tf.map_fn(map_function, elems, dtype=inputs.dtype, back_prop=True)
+
+
+def tf_dense_layer(
```

Comment:
sorry about backtracking but seeing it implemented, I see that we will use sparse version only ones. I think it is way more clear to split them